### PR TITLE
SI-7700 @unspecialized, Part Deux: Now Working.

### PR DESCRIPTION
**Pull request by retronym**
This annotation was introduced to allow us to mark methods within
a specialized trait as immune from specialization. In particular,
this is desirable for `Function1.{andThen, compose}`.

However, it seems we need to check for this in two places in the
specialization code base. The feature is now backed with a test.

Review by @VladUreche

This will introduce a binary incompatibility in the library wrt
2.11.0-M8. We'll only notice this if partest or scalacheck calls
`andThen` or `compose` on `Function1`; our PR validation build
should tell us whether that's a problem or not.

### CODE REVIEW

File: src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala [hunk 1]
```diff
@@ -306,6 +306,9 @@ abstract class SpecializeTypes extends InfoTransform with TypingTransformers {
   /** Return the specialized name of 'sym' in the given environment. It
    *  guarantees the same result regardless of the map order by sorting
    *  type variables alphabetically.
+   *
+   *  !!! Is this safe in the face of the following?
+   *    scala> trait T { def foo[A] = 0}; object O extends T { override def foo[B] = 0 }
```
Comment (VladUreche) [inline src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala:311]: Jason, what do you mean by "safe in the face of"?

**Comment by VladUreche** [comment]
LGTM, thanks @retronym! Indeed, since we have separate normalization and specialization, we need to check the annotation in both cases.

**Comment by VladUreche** [comment]
Btw, I have a stash of fixes to specialization which I did while working on miniboxing -- but they never materialized into PRs because they interact with the `abstract override` patch by @axel22. So if you set your sights on a bug, ping me, I may have a quick fix.

File: src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala (hunk 1 above)
Comment (retronym) [inline src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala:311]:
``` scala
trait T[@specialized(Int) X] {
  def foo[@specialized(Long) B, @specialized(Long, Int) A](b: B, a: A) = { println("T::foo"); b}
}

trait U extends T[Int] {
  override def foo[@specialized(Long) A, @specialized(Long, Int) B](a: A, b: B) = {println("U::foo"); a}
}

object Test extends App {
  new U {}.foo(0L, 0)
  (new U {}: T[Int]).foo(0L, 0) // !!! "T::foo"
}

```

Comment (retronym) [inline src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala:311]:
```
qbin/scalac -Xprint:specialize sandbox/test.scala && qbin/scala Test
[[syntax trees at end of                specialize]] // test.scala
package <empty> {
  abstract trait T[@specialized(scala.Int) X] extends Object {
    def /*T*/$init$(): Unit = {
      ()
    };
    def foo[@specialized(scala.Long) B, @specialized(scala.Long, scala.Int) A](b: B, a: A): B = {
      scala.this.Predef.println("T::foo");
      b
    };
    <specialized> def foo$mIJc$sp(b: Long, a: Int): Long = {
      scala.this.Predef.println("T::foo");
      b
    };
    <specialized> def foo$mJJc$sp(b: Long, a: Long): Long = {
      scala.this.Predef.println("T::foo");
      b
    }
  };
  abstract trait U extends Object with T$mcI$sp {
    def /*U*/$init$(): Unit = {
      ()
    };
    override def foo[@specialized(scala.Long) A, @specialized(scala.Long, scala.Int) B](a: A, b: B): A = {
      scala.this.Predef.println("U::foo");
      a
    };
    override <specialized> def foo$mJIc$sp(a: Long, b: Int): Long = {
      scala.this.Predef.println("U::foo");
      a
    };
    override <specialized> def foo$mJJc$sp(a: Long, b: Long): Long = {
      scala.this.Predef.println("U::foo");
      a
    }
  };
  object Test extends Object with App {
    def <init>(): Test.type = {
      Test.super.<init>();
      ()
    };
    {
  final class $anon extends Object with U {
    def <init>(): <$anon: U> = {
      $anon.super.<init>();
      ()
    }
  };
  new <$anon: U>()
}.foo$mJIc$sp(0L, 0);
    ({
  final class $anon extends Object with U {
    def <init>(): <$anon: U> = {
      $anon.super.<init>();
      ()
    }
  };
  new <$anon: U>()
}: T[Int]).foo$mIJc$sp(0L, 0)
  };
  abstract <specialized> trait T$mcI$sp extends AnyRef with T[Int]
}
```

Comment (retronym) [inline src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala:311]: Lodged: https://issues.scala-lang.org/browse/SI-8203

**Comment by retronym** [comment]
It would be great if you could publish your WIP branches and attach them to the corresponding tickets (raising new tickets if needed.) Don't hide your light under a bushel :smile:

**Comment by retronym** [comment]
Merge review by @adriaanm to be aware of the binary-compat change from the previous milestone (even though PR validation made it through.)

/cc @gkossakowski this reduces the number of methods one needs in a [SAM Function1](https://github.com/retronym/java-8-function1).

File: src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala (hunk 1 above)
Comment (VladUreche) [inline src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala:311]: Thanks for filing the bug Jason! I don't think that bug should prevent merging this PR in any way -- the issue is not related.

**Comment by VladUreche** [comment]
>  It would be great if you could publish your WIP branches and attach them to the corresponding tickets (raising new tickets if needed.) Don't hide your light under a bushel :smile:

Okay, I'll make sure I'm verbose about specialization-related issues. Let's start with https://issues.scala-lang.org/browse/SI-8204 :smiley:

**Comment by retronym** [comment]
I'm also interested in your thoughts on whether we could defining a subset of the current specialization that delivers 80% of the value but offers drastically less opportunity for bugs. For example, what if we _only_ cared about specialized `Function1` and `Function2`?

We need to do a better job and deprecated know-to-be-buggy parts of the compiler, and the advanced parts of specialization are a prime candidate.

**Comment by gkossakowski** [comment]
I'd start with deprecating the following scenario:

``` scala
class A[T @specialized]
class B[T] extends A[T]
```

Conceptually, this is hard to grasp and makes implementation difficult to maintain. I'd prefer that once you opted-in to specialized world you have to stay there when you inherit.

**Comment by VladUreche** [comment]
> I'm also interested in your thoughts on whether we could defining a subset of the current specialization that delivers 80% of the value but offers drastically less opportunity for bugs. For example, what if we only cared about specialized `Function1` and `Function2`?

That's a very hard question. Due to the [inlining problem](http://www.azulsystems.com/blog/cliff/2011-04-04-fixing-the-inlining-problem) we really need `Function1` and `Function2` to be specialized, but we might be able to stop there, by dropping (class) specialization and only keeping normalization (= method specialization). Once there, adding specializable multi-param value classes might get us far enough, without so many weak points.

Anyway, I think simplifying specialization has to do with avoiding dynamic dispatch, either through value classes, macros or type classes with normalized methods.

Btw, @gkossakowski's example is also hinting in that direction, since it exercises overriding, one of the most complex problems specialization has to address. But we can't deprecate that pattern, as it's exactly the pattern we're using for specialized closures (`$anon1$ extends Function2mcII$sp`).

/cc @dragos, @Ichoran and @non

**Comment by VladUreche** [comment]
> That's a very hard question. Due to the inlining problem we really need `Function1` and `Function2` to be specialized

Just to explain what I mean: since the function's apply methods are never inlined (due to their call sites becoming megamorphic), escape analysis can't reduce boxing at the caller-callee boundary. So we have to have a solution at the higher level for this problem, at least until Graal will know about higher-order methods. (/cc @magarciaEPFL).

**Comment by retronym** [comment]
What aspect of our patterns of `Function1`/`Function2` specialization makes them immune from the other bugs in specialization? Roughly what proportion of the implementation of class specialization would be required to support just this?

**Comment by VladUreche** [comment]
> What aspect of our patterns of `Function1`/`Function2` specialization makes them immune from the other bugs in specialization? 

They would become a special case, for which we can have a much simpler implementation.

> Roughly what proportion of the implementation of class specialization would be required to support just this?

@dragos probably knows better, but I'd say 50% at most.

File: src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala [hunk 2]
```diff
@@ -391,13 +394,16 @@ abstract class SpecializeTypes extends InfoTransform with TypingTransformers {
    *      enclosing member with the annotation.
    */
   private def needsSpecialization(env: TypeEnv, sym: Symbol): Boolean = (
-    !sym.ownerChain.exists(_ hasAnnotation UnspecializedClass) && (
+    !hasUnspecializableAnnotatation(sym) && (
```
Comment (axel22) [inline src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala]: Should be `hasUnspecializableAnnotation`, no?

Comment (VladUreche) [inline src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala]: It's a double negation -- doesn't have "not specializable" => is specializable.

Comment (retronym) [inline src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala]: He means the typo.

Comment (VladUreche) [inline src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala]: Oops, sorry, didn't notice that.

Comment (retronym) [inline src/compiler/scala/tools/nsc/transform/SpecializeTypes.scala]: Just pushed a fix for my stuttering spelling.

**Comment by retronym** [comment]
Just a typo fixed in the last commit, so I'll merge now.

**Pull request merged**